### PR TITLE
Balder/avoid plt indirection and closer packing of data

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli or @toregge PR
The first commit is a reapplication of #27646.

### CODE REVIEW

**Review by vekterli — DISMISSED** [review]
:+1:

File: vespalib/src/vespa/vespalib/datastore/compaction_strategy.h [hunk 1]
```diff
@@ -25,56 +25,53 @@ class CompactionStrategy
...
-          _max_buffers(max_buffers),
-          _active_buffers_ratio(active_buffers_ratio)
-    {
-    }
-    double getMaxDeadBytesRatio() const { return _maxDeadBytesRatio; }
-    double getMaxDeadAddressSpaceRatio() const { return _maxDeadAddressSpaceRatio; }
+          _active_buffers_ratio(active_buffers_ratio),
+          _max_buffers(max_buffers)
+    { }
+    float getMaxDeadBytesRatio() const noexcept { return _maxDeadBytesRatio; }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/datastore/compaction_strategy.h]: Good point, return as double via accessors.

File: vespalib/src/vespa/vespalib/util/growstrategy.h [hunk 2]
```diff
@@ -4,14 +4,15 @@
...
 namespace vespalib {
 
 class GrowStrategy {
 private:
-    size_t _initialCapacity;
-    size_t _minimumCapacity;
-    size_t _growDelta;
+    uint32_t _initialCapacity;
+    uint32_t _minimumCapacity;
+    uint32_t _growDelta;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/util/growstrategy.h:15]: Fixed

File: searchlib/src/vespa/searchcommon/attribute/predicate_params.h [hunk 3]
```diff
@@ -11,17 +11,16 @@ namespace search::attribute {
  */
 class PredicateParams : public PersistentPredicateParams
 {
-    double _dense_posting_list_threshold;
+    float _dense_posting_list_threshold;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchcommon/attribute/predicate_params.h:14]: This is a value with range [0.0, 1.0] so I see no problem here.

File: vespalib/src/vespa/vespalib/text/utf8.h [hunk 4]
```diff
@@ -52,20 +52,20 @@ class Utf8
...
      * @return 0, 1, 2, or 3
      **/
-    static int numContBytes(unsigned char c) {
+    static int numContBytes(unsigned char c) noexcept {
         if (c < 0x80) return 0;
         if (c > 0xC1 && c < 0xE0) return 1;
         if (c > 0xDF && c < 0xF0) return 2;
         if (c > 0xEF && c < 0xF5) return 3;
-        throwX("invalid first byte of UTF8 sequence", c);
+        return -1;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/text/utf8.h]: Good point, I did not check for all possible usages then.
I double checked all usages and saw that all was good.

**Pull request merged**